src/Bicep.Core/Emit/ExpressionConverter.cs

```diff
@@ -151,11 +152,35 @@ public LanguageExpression ConvertExpression(SyntaxBase expression)
             return (moduleSymbol, propertyAccess.PropertyName.IdentifierName);
         }
 
-        private LanguageExpression GetResourceNameExpression(ResourceSymbol resourceSymbol)
+        public LanguageExpression GetResourceNameExpression(ResourceSymbol resourceSymbol)
         {
             // this condition should have already been validated by the type checker
             var nameValueSyntax = resourceSymbol.SafeGetBodyPropertyValue(LanguageConstants.ResourceNamePropertyName) ?? throw new ArgumentException($"Expected resource syntax body to contain property 'name'");
-            return ConvertExpression(nameValueSyntax);
+
+            // For a nested resource we need to compute the name
+            var ancestors = this.context.SemanticModel.ResourceAncestors.GetAncestors(resourceSymbol);
+            if (ancestors.Length == 0)
+            {
+                return ConvertExpression(nameValueSyntax);
+            }
+
+            var segments = new List<SyntaxBase>();
+            foreach (var ancestor in ancestors)
+            {
+                var segment = ancestor.SafeGetBodyPropertyValue(LanguageConstants.ResourceNamePropertyName) ?? throw new ArgumentException($"Expected resource syntax body to contain property 'name'");
+                segments.Add(segment);
+            }
+
+            segments.Add(nameValueSyntax);
+
+            // Build an expression like '${parent.name}/${child.name}'
+            var tokens = new List<Token>();
+            tokens.Add(SyntaxFactory.CreateToken(TokenType.StringLeftPiece, ""));
+            tokens.AddRange(Enumerable.Range(0, segments.Count - 1).Select(_ => SyntaxFactory.CreateToken(TokenType.StringMiddlePiece, "/")));
+            tokens.Add(SyntaxFactory.CreateToken(TokenType.StringRightPiece, ""));
+
+            var format = new StringSyntax(tokens, segments, tokens.Select(t => t.Text));
```

Comment:
>StringSyntax [](start = 29, length = 12)

The mini AST created here hasn't gone through any of the normal validation. @anthony-c-martin, this is similar to our recent IM conversation. Do you have the same concerns here?

Reply:
I had some discussion offline with Anthony - since there isn't an IR right now the most obvious way to codegen something is with syntax nodes.

Would it make more sense to use `LanguageExpression` for this? Ultimately I need to synthesize the output for this property.

Reply:
My issue with rewriting Bicep syntax to influence the output is that you're not doing a full compiler pass over the output - you're just relying on the fact that conversion from Bicep -> JSON is mostly 1:1, but by taking a dependency on this it'll become hard to change.

Doing things with `LanguageExpression` sounds perfectly fine.

Reply:
In terms of implementation, I think it should be straightforward to copy what `GenerateUnqualifiedResourceId` does to build the format string, and pass it an enumerable of the ancestor name syntaxes piped through `ConvertExpression`.